apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+          |> put_in([:units, unit.id, :energy], current_state.units[unit.id].energy + unit.basic_skill.energy_regen)
 
         true ->
           current_state
       end
 
-    put_in(
-      new_state,
-      [unit.id, :basic_skill, :remaining_cooldown],
-      max(new_state[unit.id].basic_skill.remaining_cooldown - 1, 0)
-    )
+    # Reduce modifier remaining timers & remove expired ones
+    new_modifiers = %{
+      additives: reduce_modifier_timers(unit.modifiers.additives, unit),
+      multiplicatives: reduce_modifier_timers(unit.modifiers.multiplicatives, unit),
+      overrides: reduce_modifier_timers(unit.modifiers.overrides, unit)
+    }
 
-    # TODO: decrease_effects_remaining_steps(unit, current_state)
+    # Reduce basic skill cooldown
+    new_state
+    |> put_in(
+      [:units, unit.id, :basic_skill, :remaining_cooldown],
+      max(new_state.units[unit.id].basic_skill.remaining_cooldown - 1, 0)
+    )
+    |> put_in([:units, unit.id, :modifiers], new_modifiers)
   end
 
-  defp remove_dead_units(units),
-    do: Enum.filter(units, fn {_id, unit} -> unit.health > 0 end) |> Enum.into(%{})
+  # Reduces modifier timers and removes expired ones.
+  # Called when processing a step for a unit.
+  defp reduce_modifier_timers(modifiers, unit) do
+    Enum.reduce(modifiers, [], fn modifier, acc ->
+      case modifier.remaining_steps do
+        # Modifier is permanent
+        -1 ->
+          [modifier | acc]
+
+        # Modifier expired
+        0 ->
+          Logger.info("Modifier [#{format_modifier_name(modifier)}] expired for #{format_unit_name(unit)}.")
+
+          acc
+
+        # Modifier still going, reduce its timer by one
+        remaining ->
+          Logger.info(
+            "Modifier [#{format_modifier_name(modifier)}] remaining time reduced for #{format_unit_name(unit)}."
+          )
 
-  defp check_winner(units, step) do
-    winner =
-      cond do
-        Enum.empty?(units) -> :tie
-        Enum.all?(units, fn {_id, unit} -> unit.team == 2 end) -> :team_2
-        Enum.all?(units, fn {_id, unit} -> unit.team == 1 end) -> :team_1
-        true -> :none
+          [Map.put(modifier, :remaining_steps, remaining - 1) | acc]
       end
+    end)
+  end
 
-    case winner do
-      :none ->
-        if step == @maximum_steps,
-          do: {:halt, :timeout},
-          else: {:cont, units}
-
-      result ->
-        {:halt, result}
+  # Calculate the new state of the battle after a step passes for a skill being cast.
+  # Reduces the remaining animation and effect trigger, and casts the effects if the latter has ended.
+  defp process_step_for_skill(skill, current_state, initial_step_state) do
+    # Check if the casting unit has died
+    if Enum.any?(current_state.units, fn {unit_id, _unit} -> unit_id == skill.caster_id end) do
+      {new_skill, new_state} = process_skill_effects_trigger_value(skill, current_state, initial_step_state)
```

Comment:
i think `process_skill_effects_trigger_value` it's too short to be extracted in a different method and doesn't adds up to the readability of the code we could replace this with a `cond do` or `case skill do` something like that

Reply:
Mmh I think it's clearer when all the process_step_for_x follow the same pattern but I'll change it if you think it's not

Reply:
Nevermind, un-nesting it results in a warning because "the funciton body is nested too deep", seems like nesting is the way to go